crates/core/src/client/messages.rs

```diff
@@ -242,6 +251,156 @@ impl ToProtocol for SubscriptionUpdateMessage {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct SubscriptionRows {
+    pub table_id: TableId,
+    pub table_name: Box<str>,
+    pub table_rows: FormatSwitch<ws::TableUpdate<BsatnFormat>, ws::TableUpdate<JsonFormat>>,
```

Comment:
IMO the extent to which these format switches are plumbed throughout the code would still be unfortunate. I think it would be significantly simpler to have protocol-agnostic messages that are sent to all clients, and let each client decide how those messages will get serialized when we are actually sending the bytes.

It seems like this was done for performance reasons, specifically:
1.  For broadcasting updates, encoding immediately makes memoizing the encoding easier, and 
2. For initial queries, it saves the need to hold on to an intermediate representation).

That helps reduce the total amount of work done, but it comes at the downside of doing all of the encoding inside of transactions, which is not ideal. Memoizing updates could probably still be accomplished, but I'd have to think more about the best way to implement that.

Do you know if we have experimented with delaying the encoding?

Reply:
To my knowledge we have not experimented, except in the sense that the original implementation did a naive non-memoized version at the end of the pipeline, and we measured that to be too slow.

I am going to mark this conversation as resolved, since I don't want it to block this PR. If you would like to continue exploring, please make a ticket so we can follow up in the future.